Move GridMapEditor to bottom panel

Nodragem opened the pull request:
implements the proposal: https://github.com/godotengine/godot-proposals/issues/10405

https://github.com/user-attachments/assets/9c24679c-5cc6-4302-a8d5-6c40aed6fea6

with:
- minimum changes to the current tools and functionality of Gridmap;
- a cursor that changes color depending on the tool you are using
- I might have corrected some unexpected behaviours along the way (e.g. no tile selected when first opening)

*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot/issues/19860
- Fixes https://github.com/godotengine/godot/issues/74520
- Fixes https://github.com/godotengine/godot/issues/87487
- Fixes https://github.com/godotengine/godot/issues/92655
- Closes: https://github.com/godotengine/godot-proposals/issues/10405

Nodragem committed:
Move GridMapEditor to bottom panel

Nodragem commented:
the conflict does not seem to be super important, it's about a `#include`, but I would need to double-check which `#include` are really needed.
```
<<<<<<< moving-gridmap-panel-one-commit
#include "editor/editor_command_palette.h"
=======
#include "editor/editor_main_screen.h"
>>>>>>> master
```

ShawnHardern commented:
I think this would be a fantastic addition! 🥳 From an engine usability perspective, this change greatly improves the consistency between the placement of the GridMapEditor and the TileMap (for example). From personal experience, the difference has been a bit confusing, and this update would make the UI feel much more cohesive and intuitive. Was there originally a reason the GridMapEditor was placed on the right and not made consistent with something like TileMap?

HeadClot commented:
I am very happy that grid map is getting some much needed love :)

Nodragem commented:
I forgot to mentionned that:
- I resolved the merge conflict (I don't use the minimim width anymore),
- I cleaned the code so that it passes the automated checks,

The merge conflict made me realised that we can probably delete some of the editor settings for Gridmap which aren't relevant anymore.

smix8 commented:
In general in favor of moving it to the bottom panel. 

Not sure about all the "extras" in this PR. I tested a little and while overall it worked decent it did not take long to encounter all kinds of random visual or selection bugs. E.g. a giant version of the new selection cursor appearing at scene load out of nowhere. I had multiple random instances when playing with the toolbar options the selection wouldn't update correctly. Maybe some of those issues already existed but since there where no visuals it was hard to notice. Whatever the case this needs more testing from people.

smix8 commented:
The "big selection cursor" happens for me by just opening a scene with a GridMap node or switching to it. The GridMap node does not even need to be selected or something.

I think this is one of the main issues that should be fixed for this PR because it is something that all users will notice.

 I am personally fine with keeping the already existing GridMap bugs for other PRs, rome wasn't (re)build in a day as well.

Nodragem commented:
Hello, I had a go with the code today.

Here are the changes/fixes:
- made sure that the cursor is not visible when entering the scene tree (the giant cube issue)
- fixed the rotation of the selection when using `keep selection` and rotations during move or duplicate action 
- reintroduced the `clear rotation` action in the tool popup menu (quite useful to reset the rotation!!)
- deleted a few lines of code that where obsolete

**Before we can merge this PR**, I would like to fix a bug that was already there before but that is a little bit worse now:
- when using Gridmap, we can't use the move/rotate/scale tool of the 3D Editor to transform the Gridmap

This issue already existed, but you could use ESC to unselect the current tile in the Gridmap's palette. In this particular state of the Gridmap you could use the move/rotate/scale tool of the 3D editor. This walkaround won't work anymore, and it was not easy to find anyway.

What I would like to do is to add a `transform mode` on the left to the `selection mode`:
![image](https://github.com/user-attachments/assets/535ffb43-9a60-433e-975b-de953234847a)

It will be obvious to find for the users and in this mode, the input will be relegated to the 3D Editor, allowing the move/rotate/scale tool to work.

Nodragem commented:
Today I added the Transform Mode which can be toggled with T. This makes it easy for the user to be in a state where they can move, rotate, scale the gridmap.

**When in Transform Mode,** the Transform gizmo of the 3D Editor will be visible and QWER selects the 3D Editor tools: Node Selection, Move, Rotate and Scale.

**When NOT in Transform Mode,** the Transform gizmo of the 3D editor is hidden, and QWER selects the Gridmap tools: Selection Mode, Paint Mode, Erased Mode, etc.

Here is a demo (although we don't see which key I am typing!):

https://github.com/user-attachments/assets/1e1dd73a-8b4b-4c3c-8c73-41c24a581fd2

I also fixed the following (pre-existing) bug where the selection floor was not applying the transform of the Gridmap:
![Screenshot 2024-10-13 112155](https://github.com/user-attachments/assets/18dbe364-44c6-4833-8ca2-8e2c1564d5bc)

Nodragem commented:
Speaking with @smix8 today, we think that for this PR, I may have to drop the following feature:
- when _not_ in Transform mode in the Gridmap bottom panel, the Transform gizmo of the 3D Editor is hidden. 

This is because I edited the `Node3DEditorViewport` code to achieve that and it might delay the reviewing process.

Alternatively, @smix8 suggested that I could implement a `EditorNode3DGizmoPlugin` for Gridmap and hence show different gizmos depending on the tool mode. This would be a cleaner way to implement the different cursors, but I am not sure that it would allow me to hide the 3D Editor's Transform gizmo. I would also be a lot of refactoring work so better in another PR!

If anyone can think of an easy way to hide/show the Transform gizmo of the 3D Editor, please let me know!

Nodragem commented:
I implemented the new solution to hide/show the transform gizmo, using `set_state()`. Hence all my changes are now contained within the Gridmap module.

kitbdev commented:
Is this the same bug that was fixed?
- https://github.com/godotengine/godot/issues/19860

If so, then related:
- https://github.com/godotengine/godot/pull/97664

Nodragem commented:
> Is this the same bug that was fixed?
> 
>     * [Gridmap spatial gizmo is not responding #19860](https://github.com/godotengine/godot/issues/19860)

yes this PR solves this issue! The bottom panel adds the concept of "tool mode" so you can select between transform, paint, erase, pick and selection mode. In transform mode, the inputs are simply delegated to the 3D Editor and transform gizmo. In other modes, the transform gizmo is hidden.

Note: I don't why the videos don't work anymore :/ might need to update them.

Nodragem commented:
If I understand well, the conflict is about replacing `set_icon` with `set_button_icon`, so it should not be too complicated. I will have a look now.

smix8 reviewed (approved):
Approving the change to bottom panel. I think it is a good change and consistent with TileMap. The side panel also limited what could be reasonably added in the editor for GridMap considering how already cramped it was around the main Viewport window.

I didn't encounter (new) bugs while testing it except issues that already exists in GridMap and need to be solved in with different PRs.

PR still needs a code review from someone.

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.h:
```diff
@@ -71,11 +75,34 @@ class GridMapEditor : public VBoxContainer {
+	Button *rotate_y_button = nullptr;
+	Button *rotate_z_button = nullptr;
+
+	// toggle
+	// Button *keep_selected_button = nullptr;
+
```
```suggestion
```
Remove dead code

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
+	fill_action_button->connect(SceneStringName(pressed),
+			callable_mp(this, &GridMapEditor::_menu_option).bind(MENU_OPTION_SELECTION_FILL));
+	action_buttons->add_child(fill_action_button);
+	// viewport_shortcut_buttons.push_back(fill_action_button);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
+	move_action_button->connect(SceneStringName(pressed),
+			callable_mp(this, &GridMapEditor::_menu_option).bind(MENU_OPTION_SELECTION_CUT));
+	action_buttons->add_child(move_action_button);
+	// viewport_shortcut_buttons.push_back(move_action_button);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
+	duplicate_action_button->connect(SceneStringName(pressed),
+			callable_mp(this, &GridMapEditor::_menu_option).bind(MENU_OPTION_SELECTION_DUPLICATE));
+	action_buttons->add_child(duplicate_action_button);
+	// viewport_shortcut_buttons.push_back(duplicate_action_button);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
+	rotate_x_button->connect(SceneStringName(pressed),
+			callable_mp(this, &GridMapEditor::_menu_option).bind(MENU_OPTION_CURSOR_ROTATE_X));
+	rotation_buttons->add_child(rotate_x_button);
+	// viewport_shortcut_buttons.push_back(rotate_x_button);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
+	rotate_y_button->connect(SceneStringName(pressed),
+			callable_mp(this, &GridMapEditor::_menu_option).bind(MENU_OPTION_CURSOR_ROTATE_Y));
+	rotation_buttons->add_child(rotate_y_button);
+	// viewport_shortcut_buttons.push_back(rotate_y_button);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
+	rotate_z_button->connect(SceneStringName(pressed),
+			callable_mp(this, &GridMapEditor::_menu_option).bind(MENU_OPTION_CURSOR_ROTATE_Z));
+	rotation_buttons->add_child(rotate_z_button);
+	// viewport_shortcut_buttons.push_back(rotate_z_button);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -1275,40 +1323,187 @@ GridMapEditor::GridMapEditor() {
 	search_box = memnew(LineEdit);
-	search_box->set_h_size_flags(SIZE_EXPAND_FILL);
+	search_box->add_theme_constant_override("minimum_character_width", 10);
+	// search_box->set_h_size_flags(SIZE_EXPAND_FILL);
```
```suggestion
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.cpp:
```diff
@@ -613,13 +630,93 @@ void GridMapEditor::_do_paste() {
 	_clear_clipboard_data();
 }
 
+void GridMapEditor::_show_viewports_transform_gizmo(bool value) {
+	Dictionary new_state;
+	new_state["transform_gizmo"] = value;
```
```suggestion
void GridMapEditor::_show_viewports_transform_gizmo(bool p_value) {
	Dictionary new_state;
	new_state["transform_gizmo"] = p_value;
```

AThousandShips commented on modules/gridmap/editor/grid_map_editor_plugin.h:
```diff
@@ -208,6 +242,7 @@ class GridMapEditor : public VBoxContainer {
 	void _set_clipboard_data();
 	void _update_paste_indicator();
 	void _do_paste();
+	void _show_viewports_transform_gizmo(bool value);
```
```suggestion
	void _show_viewports_transform_gizmo(bool p_value);
```

Nodragem replied:
hello! thank you for reviewing the code. What is the fatest method to accept your changes?
 
I have a button "Commit suggestion" for each of your suggestions, but it would then split the PR into several commits, while you prefer it in a single commit.

AThousandShips replied:
You can add suggestions to a batch and commit all together, then you'll need to squash your commits into one, see [the interactive rebase](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase), or you can apply them manually

Repiteo commented:
Thanks!

geowarin commented:
@Nodragem This is a really great change, thank you for your amazing work! :smile: 

However I fear that this reintroduced a bug related to conflicting shortcuts in the editor originally fixed in #79529.

For instance, I have bound the S key to "begin scale transformation" (blender-like shortcuts).

If I'm on the gridmap and I press S to rotate a mesh, it also triggers the scaling of the whole gridmap, making a mess.

Nodragem commented:
I will have a look today! Should the fix be a new PR?

akien-mga commented:
> Should the fix be a new PR?

Yes, as this PR was merged so it can't be edited further.

Nodragem commented:
Here is the new PR: https://github.com/godotengine/godot/pull/99170
Nice catch @geowarin !

geowarin commented:
After using it a bit more, woah, what a glow up!

Here are a couple of random thoughts, for future improvements:

- I'd like the picker tool to automatically switch to paint mode after picking a mesh
- The fill tool only works when in paint mode. Whereas the intuitive thing is to go to select mode, select an area an click paint immediately, which does not work (you have to switch to select, select area, switch to paint, fill)
- I'd love the arrow keys to move to the next and previous mesh for total keyboard control :smile:

Nodragem commented:
@geowarin Thanks for the feedback and the ideas.

Would you like to leave your ideas for the next steps in this meta-proposal I've made: https://github.com/godotengine/godot-proposals/issues/10992 ?  Like that we can discuss all related ideas in one place :) thank you!
